yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,785 @@
+    public indexedUpdates: Map<IndexedTreeId, Map<bigint, IndexedTreeLeafPreimage>>,
+  ) {}
+
+  static async create(treeDb: MerkleTreeReadOperations): Promise<AvmEphemeralForest> {
+    const treeMap = new Map<MerkleTreeId, EphemeralAvmTree>();
+    for (const treeType of [MerkleTreeId.NULLIFIER_TREE, MerkleTreeId.NOTE_HASH_TREE, MerkleTreeId.PUBLIC_DATA_TREE]) {
+      const treeInfo = await treeDb.getTreeInfo(treeType);
+      const tree = await EphemeralAvmTree.create(treeInfo.size, treeInfo.depth, treeDb, treeType);
+      treeMap.set(treeType, tree);
+    }
+    return new AvmEphemeralForest(treeDb, treeMap, new Map(), new Map());
+  }
+
+  fork(): AvmEphemeralForest {
+    return new AvmEphemeralForest(
+      this.treeDb,
+      cloneDeep(this.treeMap),
+      cloneDeep(this.indexedTreeMin),
+      cloneDeep(this.indexedUpdates),
+    );
+  }
+
+  /**
+   * Gets sibling path for a leaf - if the sibling path is not found in the tree, it is fetched from the DB
+   * @param treeId - The tree to be queried for a sibling path.
+   * @param index - The index of the leaf for which a sibling path should be returned.
+   * @returns The sibling path of the leaf.
+   */
+  async getSiblingPath(treeId: MerkleTreeId, index: bigint): Promise<Fr[]> {
+    const tree = this.treeMap.get(treeId)!;
+    let path = tree.getSiblingPath(index);
+    if (path === undefined) {
+      // We dont have the sibling path in our tree - we have to get it from the DB
+      path = (await this.treeDb.getSiblingPath(treeId, index)).toFields();
+      // Since the sibling path could be outdated, we compare it with nodes in our tree
+      // if we encounter a mismatch, we replace it with the node we found in our tree.
+      for (let i = 0; i < path.length; i++) {
+        const siblingIndex = index ^ 1n;
+        const node = tree.getNode(siblingIndex, tree.depth - i);
+        if (node !== undefined) {
+          const nodeHash = tree.hashTree(node, i + 1);
+          if (!nodeHash.equals(path[i])) {
+            path[i] = nodeHash;
+          }
+        }
+        index >>= 1n;
+      }
+    }
+    return path;
+  }
+
+  /**
+   * This does the work of appending the new leaf and updating the low witness
+   * @param treeId - The tree to be queried for a sibling path.
+   * @param lowWitnessIndex - The index of the low leaf in the tree.
+   * @param lowWitness - The preimage of the low leaf.
+   * @param newLeaf - The preimage of the new leaf to be inserted.
+   * @returns The sibling path of the new leaf (i.e. the insertion path)
+   */
+  appendIndexedTree<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    lowLeafIndex: bigint,
+    lowLeafPreimage: T,
+    newLeaf: T,
+  ): Fr[] {
+    const tree = this.treeMap.get(treeId)!;
+    const newLeafHash = this.hashPreimage(newLeaf);
+    const insertIndex = tree.leafCount;
+
+    const lowLeafHash = this.hashPreimage(lowLeafPreimage);
```

Comment:
You could  just use the term "leaf" to mean "hash of preimage" as I think we generally do. And then elsewhere you make sure to always say either "preimage/leafPreimage" or "leaf" depending on what you're referring to.